cachito/web/models.py

```diff
@@ -186,6 +186,30 @@ def to_purl(self):
             # see https://github.com/package-url/purl-spec/issues/63 for further reference
             purl_name = self.name.replace("/", "%2F")
             return f"pkg:golang/{purl_name}@{self.version}"
+
+        elif self.type == "npm":
+            purl_name = self.name.replace("@", "%40")
+            match = re.match(r"(?P<protocol>[^:]+):(?://)?(?P<suffix>.+)", self.version)
+            if not match:
+                return f"pkg:npm/{purl_name}@{self.version}"
+            protocol = match.group("protocol")
+            suffix = match.group("suffix")
+            if protocol in ("github", "gitlab", "bitbucket"):  # github:namespace/name#ref
+                match_forge = re.match(
+                    r"(?P<namespace>[^/]+)/(?P<name>[^#]+)#(?P<version>.+)$", suffix
+                )
+                if not match_forge:
+                    raise ValueError(f"Could not convert version {self.version} to purl")
+                forge = match_forge.groupdict()
+                return f"pkg:{protocol}/{forge['namespace']}/{forge['name']}@{forge['version']}"
+            elif protocol in ("git", "git+http", "git+https", "git+ssh"):
+                return f"pkg:npm/{purl_name}?vcs_url={self.version}"
+            elif protocol in ("http", "https"):
```

Comment:
@mprahl do you have any examples for this case? I am not sure what to assume here (e.g., git repos, external npm repos, tarballs, etc)

Reply:
This would just be a normal URL. For example:
https://github.com/jsplumb/jsplumb/archive/2.10.2.tar.gz

Reply:
Should we use some naive heuristics (e.g., checking for file extensions) here to try to figure out what kind of resource is provided by the URL?

This would impact on what purl param we add to it. Namely, `repository_url` or `download_url`

Reply:
@athos-ribeiro if you used `download_url`, would you need to have the version as well in the PURL? Cachito is aware of the version that is in the package.json of the tarball but it is not stored in the database, so it would not be available when generating the content manifest. My concern about adding the version is that the version can be inaccurate if the dependency comes from an HTTP URL as it may have diverged significantly from upstream.

My vote is either of the following options for the package jsplumb downloaded from a tarball:
* `pkg:npm/jsplumb@https%3A%2F%2Fgithub.com%2Fjsplumb%2Fjsplumb%2Farchive%2F2.10.2.tar.gz` (assuming the URL should be URL encoded)
* `pkg:npm/jsplumb?download_url=https%3A%2F%2Fgithub.com%2Fjsplumb%2Fjsplumb%2Farchive%2F2.10.2.tar.gz`

Reply:
We do not need the version here. I will change it to `download_url` then. Would this URL ever be an alternative registry though?

Reply:
@athos-ribeiro as of now, registry dependencies only come from the official npm registry, if that changes for some reason, we can reevaluate using the `download_url` field.